### DISCUSSION TITLE
Sofar2mqtt ver 2.0

### DISCUSSION THREAD

**Discussion by Ernest-burgess** [cmcgerty/Sofar2mqtt#41] [Q&A] [answered]
Hi I am trying to use the battery_save command I have set node red to send commands on a regular interval all of the commands work apart from the battery_save command this gives me a status feed back of 512 when I and this on my calculater with 255 it gives me 0. On checking the serial output from the interface it tells me "
Message arrived on topic Sofar2mqtt/set/auto. Message: battery_save "

So it seems to be receving the msg.

If I command it to go to standby. This works and the status of the inverter changes to standby  if I then send the command battery_save the inverter seems to switch to auto and starts to discharge .

Looking at the code it has a if (messageBool)
    fnCode = SOFAR_FN_AUTO;
  else if (messageTemp =="battery_save");
BATTERYSAVE = true 

I think that it is messageBool.  That is directing it to set it to auto and not letting it set batterysave to true that is stopping it. But I don't understand what is setting   
messageBool

Any ideas ? In the mean time I will try and down load version 1 and see if that works 
Thanks. Bill

**Comment by cmcgerty** [comment]
Hi Bill,

I recently observed similar but haven’t had a chance to investigate. Battery_Save is obviously not a feature that gets used that often as we’ve not had any other reports. For me, I don’t want the battery to discharge if the EV is changing.

If you get a chance to feedback of v1 I’d be grateful (it definitely used to work!) and I’ll try to put aside some time this weekend to look at the code in v2

Colin

On 18 Nov 2022, at 23:22, Ernest-burgess ***@***.***> wrote:

﻿

Hi I am trying to use the battery_save command I have set node red to send commands on a regular interval all of the commands work apart from the battery_save command this gives me a status feed back of 512 when I and this on my calculater with 255 it gives me 0. On checking the serial output from the interface it tells me "
Message arrived on topic Sofar2mqtt/set/auto. Message: battery_save "

So it seems to be receving the msg.

If I command it to go to standby. This works and the status of the inverter changes to standby if I then send the command battery_save the inverter seems to switch to auto and starts to discharge .

Looking at the code it has a if (messageBool)
fnCode = SOFAR_FN_AUTO;
else if (messageTemp =="battery_save");
BATTERYSAVE = true

I think that it is messageBool. That is directing it to set it to auto and not letting it set batterysave to true that is stopping it. But I don't understand what is setting
messageBool

Any ideas ? In the mean time I will try and down load version 1 and see if that works
Thanks. Bill

**Reply by Ernest-burgess** [reply]
Thanks for your feed back 
I did not get chance last night to use version 1 in the end I did a new download and compile of version 2 but 
This is still giving me the same issue. It returns the status codes when sending battery_save of 512. Which if I am decoding correct 
Is 0 for success. And bit 1 set for discharge enabled.
For me this is a needed function. As I want to use it to disable the inverter if my battery voltage gets below a preset point or my soc gets below a pre set point measured on a external  lead acid battery monitoring system . If you are able to check on your system that would be a big help I may be doing something wrong at my end the string I send via node red is " battery_save "  this is sent to the topic Sofar2mqtt/set/auto. 

This is the only topic that 
Has two commands that can be sent the anther being "true" 
For auto operation. 
Is my above interpretation correct. 
Regards bill

**Comment by cmcgerty** [comment, marked as answer]
Hi Bill,

You are spot on. Line 410 needed to exclude battery_save. Try the latest upload or change line 410 in your code to….

bool  messageBool = ((messageTemp != "false") && (messageTemp != "battery_save"));

Colin

From: Ernest-burgess ***@***.***>
Sent: 18 November 2022 23:22
To: cmcgerty/Sofar2mqtt ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [cmcgerty/Sofar2mqtt] Sofar2mqtt ver 2.0 (Discussion #41)


Hi I am trying to use the battery_save command I have set node red to send commands on a regular interval all of the commands work apart from the battery_save command this gives me a status feed back of 512 when I and this on my calculater with 255 it gives me 0. On checking the serial output from the interface it tells me "
Message arrived on topic Sofar2mqtt/set/auto. Message: battery_save "

So it seems to be receving the msg.

If I command it to go to standby. This works and the status of the inverter changes to standby if I then send the command battery_save the inverter seems to switch to auto and starts to discharge .

Looking at the code it has a if (messageBool)
fnCode = SOFAR_FN_AUTO;
else if (messageTemp =="battery_save");
BATTERYSAVE = true

I think that it is messageBool. That is directing it to set it to auto and not letting it set batterysave to true that is stopping it. But I don't understand what is setting
messageBool

Any ideas ? In the mean time I will try and down load version 1 and see if that works
Thanks. Bill

**Comment by Ernest-burgess** [comment]
hi Colin  thanks for the speedy update  i have been running this code for the last few hours  and now i am getting a response to the battery_save command .
issues noted   we seem to have lost the feed back to this command on Sofar2mqtt/response/auto 
         silly me just checked my mqtt broker  looks like the feed back is now on   Sofar2mqtt/response/bsave_auto   

 when we switch into battery saver mode  the lcd updates to show this but the inverter that was in auto mode still runs in auto mode it does not switch to standby mode  and keeps discharging the battery  . I have seen it switch to standy mode but this was many min after command given .  looking at the code  it talks about wiggle room  ie dead band  i think that the inverter does a realy good job at keeping the inport to 0 watts  that we are sitting in this band and not is not seeing inport so leaves it in auto . i think that when we enter battery_saver mode  we need to force the inverter into standby   and then take the mains power readings  and continue .
(just my guess )
thanks  for the detail  what i need to change in the code  that helps me big time as i have managed to ad wifi maneger to you code so can be set up from mobile  phone ap .  if i want to change  the user details or name .  once i have it all working  can let you have a copy if you would like to look it over and make avaalble for others .
below is the serial output from my system  seen to get a few comms  errors   all the time my inverter remained in discharge mode 
20:25:44.449 -> Connecting to billsnet55
20:25:44.833 -> .
20:25:44.833 -> WiFi connected - ESP IP address: 192.168.0.99
20:25:45.218 -> Attempting MQTT connection...connected
20:25:47.372 -> Set start up mode: Auto
20:25:48.264 -> Get runstate: 3
20:25:52.279 -> Send heartbeat
20:25:59.798 -> Get runstate: 4
20:26:01.304 -> Send heartbeat
20:26:02.560 -> Timeout waiting for RS485 response.
20:26:02.560 -> Response too short
20:26:02.560 -> Bad heartbeat -2
20:26:09.811 -> Get runstate: 4
20:26:11.126 -> Send heartbeat
20:26:19.905 -> Send heartbeat
20:26:20.774 -> Timeout waiting for RS485 response.
20:26:20.774 -> Response too short
20:26:20.774 -> Bad heartbeat -2
20:26:20.774 -> Get runstate: 4
20:26:30.017 -> Send heartbeat
20:26:30.856 -> Get runstate: 4
20:26:40.073 -> Message arrived on topic: Sofar2mqtt/set/auto. Message: true
20:26:41.006 -> Send heartbeat
20:26:41.500 -> Get runstate: 4
20:26:50.140 -> Send heartbeat
20:26:51.013 -> Get runstate: 4
20:27:00.243 -> Send heartbeat
20:27:01.050 -> Get runstate: 4
20:27:10.311 -> Send heartbeat
20:27:11.148 -> Get runstate: 4
20:27:20.431 -> Send heartbeat
20:27:21.316 -> Get runstate: 4
20:27:30.492 -> Send heartbeat
20:27:31.429 -> Get runstate: 4
20:27:40.528 -> Message arrived on topic: Sofar2mqtt/set/auto. Message: true
20:27:41.915 -> Send heartbeat
20:27:41.915 -> Get runstate: 4
20:27:50.457 -> Message arrived on topic: Sofar2mqtt/set/auto. Message: battery_save
20:27:51.029 -> Send heartbeat
20:27:51.495 -> Get runstate: 4
20:27:53.710 -> Grid power: 0
20:27:53.710 -> Battery save mode: auto
20:28:00.908 -> Grid power: 0
20:28:00.908 -> Battery save mode: auto
20:28:02.169 -> Send heartbeat
20:28:02.549 -> Get runstate: 4
20:28:03.932 -> Grid power: 1
20:28:03.932 -> Battery save mode: auto
20:28:11.287 -> Grid power: 3
20:28:11.287 -> Battery save mode: auto
20:28:12.136 -> Send heartbeat
20:28:12.555 -> Get runstate: 4
20:28:14.338 -> Grid power: 65535
20:28:14.338 -> Battery save mode: auto
20:28:21.682 -> Grid power: 1
20:28:21.682 -> Battery save mode: auto
20:28:22.581 -> Send heartbeat
20:28:23.355 -> Get runstate: 4
20:28:24.779 -> Grid power: 0
20:28:24.779 -> Battery save mode: auto
20:28:31.980 -> Grid power: 0
20:28:31.980 -> Battery save mode: Timeout waiting for RS485 response.
20:28:32.810 -> Response too short
20:28:32.810 -> Send heartbeat
20:28:34.100 -> Get runstate: 4
20:28:35.435 -> Grid power: 0
20:28:35.435 -> Battery save mode: auto
20:28:42.340 -> Grid power: 65535
20:28:42.340 -> Battery save mode: auto
20:28:43.176 -> Message arrived on topic: Sofar2mqtt/set/auto. Message: battery_save
20:28:43.600 -> Send heartbeat
20:28:43.600 -> Get runstate: 4
20:28:45.393 -> Grid power: 0
20:28:45.393 -> Battery save mode: auto
20:28:52.588 -> Grid power: 65529
20:28:52.588 -> Battery save mode: auto
20:28:53.948 -> Send heartbeat
20:28:54.354 -> Get runstate: 4
20:28:55.645 -> Grid power: 1
20:28:55.645 -> Battery save mode: auto
20:29:02.869 -> Grid power: 3
20:29:02.869 -> Battery save mode: auto
20:29:03.747 -> Send heartbeat
20:29:04.193 -> Get runstate: 4
20:29:05.874 -> Grid power: 4
20:29:05.874 -> Battery save mode: auto
20:29:13.086 -> Grid power: 0
20:29:13.086 -> Battery save mode: auto
20:29:13.903 -> Send heartbeat
20:29:14.330 -> Get runstate: 4

**Reply by Ernest-burgess** [reply]
Thinking on this a bit more  when we call battery_saver mode  mabey need to check to see if we are discharging  if we are  need to call standby  so we can get true inport  power value befor we enter battry_save mode this should then put us into standby . if we are not discharging then we are good continue . just thinking on this as when we recive repeat msg to put us into battery save mode we dont want to keep going in to standby as we may be exporting and want to leave it in auto .  just a thought 
bill

**Reply by cmcgerty** [reply]
Bill,

The 10W 'wiggle room' is needed as it is always our gaol to have the grid power close to zero, whether we are charging from excess solar power to avoid sending it to the grid or supplying a load from the battery to avoid importing from the grid, grid power will end up reading just a few watts. If we reduce the wiggle room too much then the inverter ends up 'fluttering' ie. switching back and forth from auto to standby rapidly, which would be bad to it. So it's a small compromise to ensure that Sofar2mqtt won't harm anyone's inverter and I'm more than satisfied to lose 10W to the grid and still call it battery_save mode. It's avoiding things like when charging a car sucks your home battery dry that it become really useful.

With the change we've made, battery_save is now working the same as it did in v1 so i'm happy to close the issue.

**Reply by Ernest-burgess** [reply]
hi colin  i understand the 10 watt wiggle room  infact i think that is quite tight .   it is the case that when you switch into battery saver mode  if the inverter is allready discharging the battery  and the power is 0  then it leaves the inverter in auto mode  ie discharging the battery  it does not seem to switch to standby untill it sees the inport go over the 10 w wiggel room .  so as long as the inverter keeps discharging the battery and keeps the power within 10 watts . it wont change state .. the inverter could be feeding up to 3kw  and still keep power at 0 .
if system is in auto when we switch to battery_save 
on entering the save_battery routeen we need to check the discharge status and if we are discharging the battery  we need to switch to standby  as first task . then allow the routeen to make the decision when to switch to auto . 
when it sees export .  
in the list of feed back you can see the command to go to battery_save was given at  27m:50s it was still in auto 
mode discharging at  29m:13s  , still keeping power in the dead band wiggel room .  unless some load in my house switched on  that would get the  inport to excede the inverter output i cant see it changing ..
sorry i did not make myself clear on this 
bill

**Reply by Ernest-burgess** [reply]
just a point of intrest  i am running  ver 3 software on my Sofar inverter me3000sp  
on this inverter there is a running mode called skimming mode that does the same as the battery_saver 
i cant see any ref to this in the register lists that i can find to see if we could switch the inverter into this mode 
is this a new mode or has it been on the preceding versions of inverter software ?

**Reply by Ernest-burgess** [reply]
hi Colin  i have made this change to the code  to solve the problem i had above of the unit not switching to standby when discharging the battery in auto mode  when the switch to battery_saver is made .

line 468  
old line  if(p < 65535/2 || p > 65525)   // added below to make inverter change to standby if in discharge mode 
  new line   if((p < 65535/2 || p > 65525)  &&  !INVERTER_RUNNINGSTATE == discharging)   // bill  added bit code to force standby mode if battery is discharging first time in to battery saver mode 

i have tested it on my system and it switches the inverter from discharge   to standby  in a few seconds after the battery save cmd is recived .     on my system with the me3000sp 

if i have understood the progran correctly this will use the origonal test  if(p < 65535/2 || p > 65525)     to control auto standby  selection 
as long as we are not discharging   , if we are discharging  it will use the  &&  !INVERTER_RUNNINGSTATE   part to select the standby mode 

regards bill

**Comment by double1968** [comment]
Hi,
Battery-Save also doesn't work on my inverter.
I reported a while ago. [battery save](https://github.com/cmcgerty/Sofar2mqtt/discussions/12)
Maybe it's an inverter firmware problem
My inverter is a ZCS Azzuro (SOFAR) HYD4000 ZSS fw. 2.70

**Comment by cmcgerty** [comment]
Hi @double1968 

Battery_save should not be inverter model related as it’s a feature of the Sofar2mqtt firmware. It is just sending either auto or standby to the inverter to allow it charge when there is power coming in but switch to standby if it starts to flow out.

Try the latest update as it should work better now.

Colin

**Reply by double1968** [reply]
Hi @cmcgerty I've been running sofar2mqtt 2.0b1 since it was released.  Unfortunately battery save does not work.

**Reply by cmcgerty** [reply]
Quite so, sorry, I should have been clearer. It was broken in 2.0 and fixed in 2.0.1.

**Comment by cmcgerty** [comment]
I don’t think that’s new. There has always been a battery saving mode you can access when you manually program the inverter but it’s not available if you use passive mode, as we do for Sofar2mqtt.

**Reply by Ernest-burgess** [reply]
hi Colin i have made this change to the code to solve the problem i had above of the unit not switching to standby when discharging the battery in auto mode when the switch to battery_saver is made .

line 468
old line if(p < 65535/2 || p > 65525) // added below to make inverter change to standby if in discharge mode
new line if((p < 65535/2 || p > 65525) && !INVERTER_RUNNINGSTATE == discharging) // bill added bit code to force standby mode if battery is discharging first time in to battery saver mode

i have tested it on my system and it switches the inverter from discharge to standby in a few seconds after the battery save cmd is recived . on my system with the me3000sp

if i have understood the progran correctly this will use the origonal test if(p < 65535/2 || p > 65525) to control auto standby selection
as long as we are not discharging , if we are discharging it will use the && !INVERTER_RUNNINGSTATE part to select the standby mode

regards bill

**Reply by cmcgerty** [reply]
Thanks for this Bill. I’m glad it’s working for you. I’ll take a look over the weekend if time allows.

**Reply by Ernest-burgess** [reply]
opps  i got it a bit wrong   "! "did not work as i expected  but it did compile  had to change the line too
new line if((p < 65535/2 || p > 65525) &&  INVERTER_RUNNINGSTATE != discharging) // bill added bit code to force standby mode if battery is discharging first time in to battery saver mode

had some solar gen this morning  and it now seems to work as i expect

**Reply by cmcgerty** [reply]
For sake of my brain, I've put brackets around the second half as well. Otherwise, a really nice solution. Thank you Bill. Note that this is not only run when you first enter battery_save mode. The whole batterySave subroutine is called at the end of the main loop every 100 milliseconds and this line is evaluated each time the 3 second BATTERYSAVE_INTERVAL timer passes while BATTERYSAVE is true. But since the result of the first time around can only be auto or charging, you are right to say that it only has an impact the first time it is called.

**Comment by Ernest-burgess** [comment]
Thanks for the update.  Seems to do the trick. Have tested it today on both charge and discharge as we had some good sun . Seems to work .
Having a bad day with my aurdino compiler.   Came to use it the other day and it just plain refused to load . Did a fresh install and decided to try the last releases . Had all sorts of issues trying to get programs to compile . Just about got it sorted now. But found that the code that I had working with the WiFi master .
Will compile but gives rs485 errors all the time .  So I have started again with a fully working version 2  of sofar2-Mqtt  and now putting the WiFi master back into it


Regards. Bill

On 26 Nov 2022, at 16:17, cmcg ***@***.***> wrote:

﻿

For sake of my brain, I've put brackets around the second half as well. Otherwise, a really nice solution. Thank you Bill. Note that this is not only run when you first enter battery_save mode. The whole batterySave subroutine is called at the end of the main loop every 100 milliseconds and this line is evaluated each time the 3 second BATTERYSAVE_INTERVAL timer passes while BATTERYSAVE is true. But since the result of the first time around can only be auto or charging, you are right to say that it only has an impact the first time it is called.
